lib/ddtrace/profiling/pprof/builder.rb

```diff
@@ -0,0 +1,181 @@
+require 'ddtrace/profiling/pprof/message_set'
+require 'ddtrace/profiling/pprof/pprof_pb'
+require 'ddtrace/profiling/pprof/string_table'
+
+module Datadog
+  module Profiling
+    module Pprof
+      # Generic profile building behavior
+      class Builder
+        DESC_FRAME_OMITTED = 'frame omitted'.freeze
+        DESC_FRAMES_OMITTED = 'frames omitted'.freeze
+        VALUE_TYPE_WALL = 'wall'.freeze
+        VALUE_UNIT_NANOSECONDS = 'nanoseconds'.freeze
+
+        attr_reader \
+          :events,
+          :functions,
+          :locations,
+          :sample_types,
+          :samples,
+          :string_table
+
+        def initialize(events)
+          @events = events
+          @profile = nil
+          @sample_types = []
+          @samples = []
+          @mappings = []
+          @locations = MessageSet.new
+          @functions = MessageSet.new
+          @string_table = StringTable.new
```

Comment:
IIUC you made those local to a builder, which make them local to an event type.
That should be local to a `Profile`. Even when it will handle several types of events, you can re-use the `locations`, `functions`, and the `string_table`.